### ISSUE TITLE
Inconsistency in implementation of _LRScheduler 

### ISSUE THREAD

**Issue by Erotemic** [pytorch/pytorch#8837]
I've noticed an odd behavior when attempting to write my own scheduler based on `torch.optim.lr_scheduler._LRScheduler`. 

If you write a custom get_lr() to work based on self.last_epoch, its impossible to differentiate the 
0th and the 1st epoch. 

Here is a minimal working example:

```python

def mwe():
    # Assuming optimizer has two groups.
    import torch.optim.lr_scheduler
    import netharn as nh
    model = nh.models.ToyNet2d()
    optimizer = torch.optim.SGD(model.parameters(), lr=10)

    class DummySchedule(torch.optim.lr_scheduler._LRScheduler):
        def get_lr(self):
            print('Set LR based on self.last_epoch = {!r}'.format(self.last_epoch))
            self._current_lr = self.last_epoch
            return [self.last_epoch]

    # Initialize the optimizer with epoch 0's LR
    # self = torch.optim.lr_scheduler.LambdaLR(optimizer, lambda x: x)
    self = DummySchedule(optimizer)
    for epoch in range(3):
        print('------')
        print('Run epoch = {!r}'.format(epoch))
        # Pretend we run epoch 0
        print('Training with self._current_lr = {!r}'.format(self._current_lr))
        # Pretend epoch 0 has finished, so step the scheduler.
        # self.step(epoch=epoch)
        self.step()
```

This results in the output

```
Set LR based on self.last_epoch = 0
------
Run epoch = 0
Training with self._current_lr = 0
Set LR based on self.last_epoch = 0
------
Run epoch = 1
Training with self._current_lr = 0
Set LR based on self.last_epoch = 1
------
Run epoch = 2
Training with self._current_lr = 1
Set LR based on self.last_epoch = 2
```

You can see the last epoch is asked to set the learning rate based on the last epoch being 0 twice. This LRScheduler class takes last_epoch as an argument, so it knows how to set the LR for the previous epoch. By default last_epoch=-1, because the first epoch is 0 and no epoch has run yet. On construction it then calls `step` with `last_epoch + 1`, which means the step function sets the learning rate for epoch 0. Then last_epoch is reset to -1 immediately after, so the next call to step also sets the learning rate for epoch 0. 

A fix would simply remove the + 1 from `self.step(last_epoch + 1)`, but this might break existing implementations of `get_lr()` which wouldn't expect `self.last_epoch` being set to a negative number. 

I think a more intuitive implementation of this class might track the current epoch rather than the previous one. This would be a backwards incompatible change, but I think it would improve the overall quality of torch. I'm willing to give a re-implementation a shot if it sounds like a good idea to the maintainers.
